### DISCUSSION TITLE
Manual revalidate ISG would have support for I18N and dynamic routes

### DISCUSSION THREAD

**Discussion by MetinArslanturk** [vercel/next.js#34634] [Ideas]
### Describe the feature you'd like to request

Hi NextJS team.

Firstly thanks for on-demand (manually) revalidation support for statically pre-rendered (getStaticProps) pages. You guys already know that we wanted and waited for this feature in hard way really as community.

After release of latest version I tried it immediately with `res.unstable_revalidate ` api. It works in smooth way. But I realized that it works only for provided url directly. I mean let's say I have a blog and I am supporting `en` and `es` languages.

I have `posts/[id]` page structure. Let's say on this post page I am also fetching and showing my total posts count from an external API.
When I add a new post I want to update all old ones with a web-hook. When I trigger revalidate as like `await res.unstable_revalidate('/en/posts/12')` , it revalidates for only post page of with ID 12 and only English version.

It would be really great if I can revalidate for all versions of that page. The I18N part is not hard. If I am supporting for example 4 languages I can loop them and trigger 4 different `res.unstable_revalidate` requests. It is OK.

But the dynamic urls are not that easy. Because you know if I use ISG I am providing some URLs with getStaticPaths as initially but some URLs being prepared on runtime with ISG. I can't know which pages generated on runtime so I can not trigger revalidate request for that pages.

### Describe the solution you'd like

There can be a custom url pattern in `res.unstable_revalidate` API for dynamic url pages.
For example: `await res.unstable_revalidate('/en/posts/*')`  The asterisk character means purge (revalidate) all routes based on `/en/posts/` .

Also it can take an options argument and it includes a flag for purging all I18N routes for that url.

i.e.: `await res.unstable_revalidate('/posts/*', {revalidateAllI18NVersions: true})` 

Now it will revalidate all I18N versions (en, es) and all paths (/posts/1, /posts/2, /posts/3..... all getStaticPaths + generated paths on runtime with ISG).

### Describe alternatives you've considered

Maybe you guys can expose `incrementalCache` to `req.context` on API handler. So, we can revalidate some pages manually in two steps:

1. Remove static generated pages from .next/server/pages
2. Purge `incrementalCache` for related urls.

I am currently doing this but this requires creating a custom server and it makes more complex to handling things. (You can check from here: https://github.com/MetinArslanturk/microfrontend-nextjs-ssr/blob/i18n-integration/next-host/scripts/utils.js ) Handling from API routes would be easier.

Also handling it with directly the new `unstable_revalidate `API would be much much easier.

If you guys check example repo (in i18n-integration branch), You will see why we need this. We are using dynamic microfrontend pieces on pages. When a microfrontend app has new deployment we want to revalidate all related pages.

Thanks from now,
Metin

**Comment by balazsorban44** [comment]
This is likely a duplicate of https://github.com/vercel/next.js/discussions/34567

**Comment by MetinArslanturk** [comment]
Not really. He means accepting array as an argument to this new API. I mean a shortcut to revalidating pages which created with ISG (getStaticPaths). I mean, I don't even have an array of my runtime generated pages. `(/posts/1, /posts/2, /posts/3..... all getStaticPaths + generated paths on runtime with ISG)` . Check this, the second part is unknown. So I advised a shortcut like asterisk.
